### PR TITLE
Extended GCD and Inverses

### LINKED ISSUE #10: Implement Extended GCD algorithm
In order to advance quickly with prototyping we can we the basic extended GCD algorithm to compute inverses, which is enough to implement HE schemes.

- [x] Basic Extended GCD algorithm.
- [x] Implement polynomial inverses in the ring.
- [x] Basic optimizations (low hanging fruits). 
- [x] Documentation.

### PR DESCRIPTION
**Pull request by emmorais**
This PR adds the Extended GCD for cyclotomic polynomials. Inverses can be computed using this primitive. For our parameters choices we have high probability of finding an invertible element for relatively small randomly sampled polynomials. Therefore we don't expect to sample multiple times before finding an invertible element. This is compatible  with experiments carried out in the C++ [repo](https://github.com/Inversed-Tech/hom-ham/tree/main/si-ntru). 

This is efficient enough to be used for key generation of the YASHE construction. 

Closes #10

### CODE REVIEW

File: eyelid-match-ops/src/primitives/poly/test/inv.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+//! Tests for polynomial inverse.
+
+use super::super::*;
+use rand::Rng;
+
+/// This sampling is similar to what will be necessary for YASHE KeyGen
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/test/inv.rs:6]: If this sampling function is going to be used in production key generation, can it move to `poly/modular_poly/inverse.rs`?

(If it’s going to be used in production, it doesn’t need to be benchmark and test only.)

File: eyelid-match-ops/src/primitives/poly/test/inv.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/// TODO: generate Gaussian distribution instead of "uniform"
+pub fn sample() -> Poly {
+    let mut rng = rand::thread_rng();
+    let mut res = Poly::zero();
+    let max_coeff = 8;
+    let t = 2;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/test/inv.rs:12]: Should these parameters be named constants or function arguments?

File: eyelid-match-ops/src/primitives/poly.rs [hunk 3]
```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
...
+    let divisor = ri_prev.clone();
+    // FIXME: if b is not invertible mod a the assert is not going to pass.
+    // Since the test is probabilistic, we have a small chance of failure.
+    // It fails after a small number of executions.
+    // Sometimes it fails with a different error, meaning there is another
+    // problem happening.
+    // It occasionally fails when we do the following:
+    // export RUSTFLAGS="-D warnings --cfg tiny_poly"
+    // cargo test --all-targets -- --nocapture inv
+    debug_assert!(divisor.degree() == 0);
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:235]: If this assert is necessary for the method to return correct results, and it’s possible for the assert to fail on some random or user inputs, then we should return an error instead.

Assertions that prevent incorrect results can’t be `debug_assert!()`, because then we could get incorrect results in production. If an incorrect result is used in key generation, that is a security issue.

And they can’t be `assert!()` either, because panicking in production is a reliability bug.

(Assertions are only for things we think are always true. They help us find bugs in new or changed code, where what we thought was wrong.)

File: eyelid-match-ops/src/primitives/poly.rs [hunk 4]
```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
...
+    (x_cur, x_prev) = update_diophantine(x_prev, x_cur, q.clone().into());
+    // y_cur = y_prev - q.y_cur
+    (y_cur, y_prev) = update_diophantine(y_prev, y_cur, q.clone().into());
+    // loop until ri_cur = 0
+    while !(ri_cur.is_zero()) {
+        let ri_aux = ri_cur.clone();
+        // TODO: q is just a monomial, then we can optimize the next computation
+        (q, ri_cur) = DenseOrSparsePolynomial::from(ri_prev.clone())
+            .divide_with_q_and_r(&ri_cur.into())
+            .expect("loop divisor is not zero");
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:218]: This explains how know the `expect()` will never panic.

```suggestion
            .expect("just checked that the loop divisor is not zero");
```

File: eyelid-match-ops/src/primitives/poly.rs [hunk 5]
```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
...
+    // next:     x1=0, y1=1, r1=b
+    let mut x_cur = Poly::zero();
+    let mut y_cur = Poly::one();
+    let ri_cur = b.clone();
+
+    let ri_aux = ri_cur.clone();
+    // TODO: q is just a monomial, then we can optimize the next computation
+    let (mut q, mut ri_cur) = DenseOrSparsePolynomial::from(ri_prev.clone())
+        .divide_with_q_and_r(&ri_cur.into())
+        .expect("init divisor is not zero");
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:206]: This will panic on `inverse(Poly::zero())`. Instead, return an error.

(See my detailed explanation below.)

File: eyelid-match-ops/src/primitives/poly/test/inv.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        let coeff_rand = rng.gen_range(1..max_coeff);
+        res[i] = Coeff::from(t * coeff_rand);
+    }
+    res[0] += Coeff::one();
+    res
+}
+
+#[test]
+fn test_inverse() {
+    let f = sample();
```
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: Maybe we could consider using [proptests](https://proptest-rs.github.io/proptest/intro.html).

Very nice suggestions!

File: eyelid-match-ops/src/primitives/poly.rs [hunk 7]
```diff
@@ -33,6 +36,24 @@ pub use cyclotomic_mul as mul_poly;
 /// Minimum degree for recursive Karatsuba calls
 pub const MIN_KARATSUBA_REC_DEGREE: usize = 32; // TODO: fine tune
 
+/// Returns a Boolean indicating if the input is equal or not to the additive
+/// identity in the polynomial ring
+pub fn is_zero_poly(a: Poly) -> bool {
```
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs:41]: yes, old functions indeed.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 8]
```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
 
     (al, ar)
 }
+
+/// Returns the inverse in the cyclotomic ring, if it exists.
```
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs:161]: I like the idea of granularity in general. I have the impression `poly.rs` has a big potential to grow, then I totally agree that it is a good example where splitting into smaller pieces would help.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 1 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:6]: Yes, I was expecting to move it in the near future, then it makes sense to `inverse.rs` now.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 2 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:12]: I just wanted to restrict the size of coefficients, because it influences probability of finding the inverse (smaller is better, i.e. more likely to find it). For the final version of this sampling, it would be good to have named constants, but they will change a bit, since the distribution will be Gaussian instead of Uniform. And maybe some parameter will be not a constant, but an attribute of the scheme (dynamic), like for example `t`.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 9]
```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
...
+    // cargo test --all-targets -- --nocapture inv
+    debug_assert!(divisor.degree() == 0);
+    let divisor_inv = divisor[0].inverse().unwrap();
+    // y_cur / ri_prev
+    let mut final_result = y_prev.clone();
+    for i in 0..=y_prev.degree() {
+        final_result[i] = final_result[i].mul(divisor_inv);
+    }
+    final_result
+}
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:243]: @bgillesp some maths for you

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: > Maybe we could consider using [proptests](https://proptest-rs.github.io/proptest/intro.html).

Sounds good, did you want to open a ticket?

**Review by teor2345 — COMMENTED** [review]
If PR #47 merges before this PR, we'll need to make some changes here:
- genric `Poly` `impl` and `fn` need an added `<const MAX_POLY_DEGREE: usize>`
- generic `Poly` types become `Poly<MAX_POLY_DEGREE>`
- concrete `Poly` types become `Poly<FULL_RES_POLY_DEGREE>`
- concrete poly degrees become `FULL_RES_POLY_DEGREE`

If you rebase on #47, the compiler will tell you what's missing. But this can be done last, just before merging.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: [Here](https://github.com/Inversed-Tech/eyelid/issues/55) it is.

File: eyelid-match-ops/src/primitives/poly.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: Removed

File: eyelid-match-ops/src/primitives/poly.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs:143]: The multiplication move is in https://github.com/Inversed-Tech/eyelid/pull/52

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: Thanks! I made a comment on that ticket about how `proptest` struggles to find rare edge cases.

Are you ok with adding the list of tests above to this PR?
Let me know if you need any help.

**Comment by teor2345** [comment]
> Cyclotomic inverse: polynomial/1 relatively small random poly of degree N

I already added "inv" to the important benchmarks list in PR #57, so this is appearing as expected.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: Done

File: eyelid-match-ops/src/primitives/poly.rs (hunk 3 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: Replaced by an error. Also removed the error treatment from inverse() function and the unnecessary check (now happening at the unit test). 

Since we have opened an [issue](https://github.com/Inversed-Tech/eyelid/issues/55) for better testing it, I think this solves the problem for now. Therefore I also removed the FIXME comment. Do you agree?

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:23]: I don’t think these tests can be done with proptests:
- [ ] Test standard known inverses, for example, that -1 and 1 are their own inverses, and 0 doesn’t have an inverse.
- [ ] If there’s a polynomial we know the inverse of, that would also be useful to test. If not, we could generate a test vector and its inverse, and check that.

So they either need a separate ticket, or they could be added to this PR. I don’t mind merging this PR without those tests, but I’d like to have tests merged soon after PR, to avoid bugs.

As part of the fixed test vectors, it might be useful to test both representations of -1 (`COEFF_MODULUS - 1` and `X^N`), as well as both representations of zero.

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 2 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs:12]: Do you think this is a blocker for merging this PR, or can it go in a ticket?

File: eyelid-match-ops/src/primitives/poly.rs (hunk 5 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs:374]: Did you need help fixing this?

File: eyelid-match-ops/src/primitives/poly/test.rs [hunk 10]
```diff
@@ -5,3 +5,6 @@ pub mod gen;
 
 #[cfg(test)]
 pub mod mul;
+
+#[cfg(any(test, feature = "benchmark"))]
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/test.rs:9]: When `sample()` is moved, this can become:
```suggestion
#[cfg(test)]
```

File: eyelid-match-ops/src/primitives/poly.rs [hunk 11]
```diff
@@ -326,3 +330,76 @@ pub fn poly_split_half(a: &Poly, chunk: usize) -> (Poly, Poly) {
...
     (remainder, quotient)
 }
+
+/// Returns the inverse in the cyclotomic ring, if it exists.
+/// Otherwise, returns an error.
+pub fn inverse(a: &Poly) -> Result<Poly, String> {
+    let mut mod_pol = Poly::zero();
+
+    // TODO: don't recompute the modulus here
+    mod_pol[MAX_POLY_DEGREE] = Coeff::one();
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:340]: There’s a function for this now, I’ll make these changes today as part of updating this code for PR #40.

File: eyelid-match-ops/src/primitives/poly.rs (hunk 5 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs:374]: Rust has convenience methods and operators for this kind of result handling:
- `Option` to `Result`: https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else
- Only executing error code when needed: https://doc.rust-lang.org/book/ch13-01-closures.html
- Returning an error `Result`: https://doc.rust-lang.org/std/result/#the-question-mark-operator-

```suggestion
        .ok_or_else(|| "init divisor is not zero".to_string())?;
```

File: eyelid-match-ops/src/primitives/poly.rs (hunk 11 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs]: Done!

File: eyelid-match-ops/src/primitives/poly.rs (hunk 5 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs]: It turns out this whole block of code is a duplicate of the code in the loop. After I removed the duplicate code, this change isn't needed any more.

(I also needed to implement a missing `Poly::divide_with_q_and_r()` method.)

File: eyelid-match-ops/src/primitives/poly.rs [hunk 12]
```diff
@@ -326,3 +330,76 @@ pub fn poly_split_half(a: &Poly, chunk: usize) -> (Poly, Poly) {
...
+
+/// Helps to calculate the equation `cur = prev - q.cur`.
+fn update_diophantine(prev: Poly, cur: Poly, q: Poly) -> (Poly, Poly) {
+    let aux = cur.clone();
+    let mul_res = cur.mul(&q.clone());
+    let new_cur = prev.sub(&mul_res).clone();
+    let new_prev = aux.clone();
+
+    (new_cur, new_prev)
+}
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs]: After implementing some missing operators, this can be simplified, and tweaked to re-use most of the supplied polynomials.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 13]
```diff
@@ -27,5 +29,88 @@ pub mod modular_poly;
 #[cfg(any(test, feature = "benchmark"))]
 pub mod test;
 
-// Do not add code here.
-// Add functions or trait impls to modular_poly/*.rs and inherent method impls to modular_poly.rs.
+/// Returns the monic inverse of `a` in the cyclotomic ring, if it exists.
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:32]:
```suggestion
// Do not add code here.
// Add functions or trait impls to modular_poly/*.rs and inherent method impls to modular_poly.rs.

/// Returns the monic inverse of `a` in the cyclotomic ring, if it exists.
```

**Review by teor2345 — APPROVED** [review]
I fixed up the code conflicts after my recent PRs, and simplified some of the code.

As part of that, I implemented some missing operators and methods on `Poly`.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 14]
```diff
@@ -29,3 +31,89 @@ pub mod test;
...
+        Err("Non-invertible polynomial".to_string())
+    } else {
+        // Reduce to a monic polynomial.
+        let mut inv: Poly<MAX_POLY_DEGREE> = y;
+        let divisor_inv: Coeff = d[0].inverse().expect("just checked for zero");
+
+        inv *= divisor_inv;
+
+        Ok(inv)
+    }
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:58]: Is `monic` the correct term here?
https://en.wikipedia.org/wiki/Polynomial_greatest_common_divisor#General_definition

If we want the inverse to be monic, should we be multiplying by the inverse of the coefficient of the degree?
That would be `y[y.degree()].inverse()`, instead of `d[0].inverse()`.

File: eyelid-match-ops/src/primitives/poly/modular_poly/trivial.rs [hunk 15]
```diff
@@ -151,7 +151,7 @@ impl<const MAX_POLY_DEGREE: usize> SubAssign<&Poly<MAX_POLY_DEGREE>> for Poly<MA
     }
 }
 
-// Multiplying by a scalar can't increase the degree, so it is trivial.
+// Multiplying by a polynomial is a trivial wrapper can't increase the degree, so it is trivial.
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/modular_poly/trivial.rs:154]:
```suggestion
// Multiplying by a scalar can't increase the degree, so it is trivial.
```

File: eyelid-match-ops/src/primitives/poly.rs (hunk 14 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/poly.rs:58]: @bgillesp this might be a maths question for you?

File: eyelid-match-ops/src/primitives/poly.rs [hunk 16]
```diff
@@ -29,3 +31,89 @@ pub mod test;
...
+        ri_prev = ri_aux;
+
+        // x_cur = x_prev - q.x_cur
+        (x_cur, x_prev) = update_diophantine(x_prev, x_cur, &q);
+        // y_cur = y_prev - q.y_cur
+        (y_cur, y_prev) = update_diophantine(y_prev, y_cur, &q);
+    }
+
+    (x_prev, y_prev, ri_prev)
+}
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:119]: @bgillesp some maths for you to check

**Comment by teor2345** [comment]
> # Benchmark
> Benchmark suite	Current: [2915b72](https://github.com/Inversed-Tech/eyelid/commit/2915b723f116564f32ee489b76140aa96d6c6eaf)	Previous: [9c62aec](https://github.com/Inversed-Tech/eyelid/commit/9c62aecad648cc8922d31668078bdfd91532655a)	Ratio
> `Full iris match: plaintext/Random bits`	`965615` ns/iter (`± 8218`)	`980280` ns/iter (`± 9526`)	`0.99`
> `Cyclotomic multiplication: polynomial/2 random polys of degree N`	`70567255` ns/iter (`± 43963`)	`66003408` ns/iter (`± 181589`)	`1.07`
> `Recursive Karatsuba multiplication: polynomial/2 random polys of degree N`	`9946048` ns/iter (`± 20919`)		
> `Flat Karatsuba multiplication: polynomial/2 random polys of degree N`	`36023771` ns/iter (`± 69736`)	`33642000` ns/iter (`± 82031`)	`1.07`
> `Cyclotomic inverse: polynomial/1 relatively small random poly of degree N`	`191523868` ns/iter (`± 220543`)		
> This comment was automatically generated by [workflow](https://github.com/Inversed-Tech/eyelid/actions?query=workflow%3AMonitor%20Benchmark%20Changes) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).

Looks like these changes saved about 5ms / 196ms.

**Comment by teor2345** [comment]
I fixed the tests for tiny polynomials, because they fail occasionally:
https://github.com/Inversed-Tech/eyelid/actions/runs/8779228223/job/24086917262?pr=48

File: eyelid-match-ops/src/primitives/poly.rs (hunk 14 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: Here we compute a primitive polynomial by dividing by the content. I fixed the terminology and added a reference to the description of the algorithm.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 17]
```diff
@@ -326,3 +330,76 @@ pub fn poly_split_half(a: &Poly, chunk: usize) -> (Poly, Poly) {
...
+    let mul_res = cur.mul(&q.clone());
+    let new_cur = prev.sub(&mul_res).clone();
+    let new_prev = aux.clone();
+
+    (new_cur, new_prev)
+}
+
+/// Returns polynomials x, y, d such that a.x + b.y = d.
+/// When d=1 we have that x is the multiplicative inverse of a.
+pub fn extended_gcd(a: &Poly, b: &Poly) -> Result<Poly, String> {
```
[the remark this answers is not in the document]
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly.rs]: I added more information to the documentation to make it clearer.

File: eyelid-match-ops/benches/match-ops.rs [hunk 18]
```diff
@@ -213,3 +222,23 @@ pub fn bench_mod_poly_ark(settings: &mut Criterion) {
         },
     );
 }
+
+/// Run [`poly::karatsuba_mul()`] as a Criterion benchmark with random data.
```
Comment (naure) [inline eyelid-match-ops/benches/match-ops.rs:226]: "karatsuba_mul" -> wrong name?

File: eyelid-match-ops/benches/match-ops.rs [hunk 19]
```diff
@@ -213,3 +222,23 @@ pub fn bench_mod_poly_ark(settings: &mut Criterion) {
...
+pub fn bench_inv(settings: &mut Criterion) {
+    // Setup: generate random cyclotomic polynomials
+    let p = sample::<FULL_RES_POLY_DEGREE>();
```
Comment (naure) [inline eyelid-match-ops/benches/match-ops.rs:229]: There might be some variability of timing introduced from sampling. Ideally this benchmark would use a deterministic RNG with a fixed seed.

It’s common for functions like `sample` to take `rng` as a parameter, and let each test decide which RNG to use.

**Review by naure — APPROVED** [review]

File: eyelid-match-ops/benches/match-ops.rs (hunk 19 above)
Comment (emmorais) [reply eyelid-match-ops/benches/match-ops.rs]: Done

File: eyelid-match-ops/src/primitives/poly/test/inv.rs (hunk 2 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/poly/test/inv.rs]: No, this is not a blocker. The sampling will be important during the implementation of YASHE, and we have a ticket for it. Will describe those sampling methods in this ticket. 

For the inverse, the only thing we need is to increase the likelihood of finding an invertible element, so this is good enough.

File: eyelid-match-ops/src/primitives/poly.rs [hunk 20]
```diff
@@ -29,3 +32,107 @@ pub mod test;
...
+
+    let mut q: Poly<MAX_POLY_DEGREE>;
+
+    // loop until ri_cur = 0
+    while !(ri_cur.is_zero()) {
+        let ri_aux = ri_cur.clone();
+        // TODO: q is just a monomial, then we can optimize the next computation
+        (q, ri_cur) = ri_prev
+            .divide_with_q_and_r(&ri_cur)
+            .expect("just checked that the loop divisor is not zero");
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:114]:
```suggestion
        (q, ri_cur) = ri_prev
            .divide_with_q_and_r(&ri_cur)
            .expect("just checked that the loop divisor is not zero");
        // Sometimes divide_with_q_and_r() might be returning a non-canonical polynomial
        ri_cur.truncate_to_canonical_form();
```

File: eyelid-match-ops/src/primitives/poly.rs [hunk 21]
```diff
@@ -29,3 +32,107 @@ pub mod test;
...
+    // init with x0=1, y0=0, r0=a
+    let mut x_prev: Poly<MAX_POLY_DEGREE> = Poly::one();
+    let mut y_prev = Poly::zero();
+    let mut ri_prev = a.clone();
+    // next:     x1=0, y1=1, r1=b
+    let mut x_cur = Poly::zero();
+    let mut y_cur = Poly::one();
+    let mut ri_cur = b.clone();
+
+    let mut q: Poly<MAX_POLY_DEGREE>;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:106]:
```suggestion
    let mut ri_cur = b.clone();

    let mut q: Poly<MAX_POLY_DEGREE>;

    // Sometimes the inputs can be non-canonical.
    ri_cur.truncate_to_canonical_form();

```

File: eyelid-match-ops/src/primitives/poly.rs [hunk 22]
```diff
@@ -29,3 +32,113 @@ pub mod test;
...
+
+    // Sometimes the inputs can be non-canonical.
+    ri_cur.truncate_to_canonical_form();
+
+
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:111]:
```suggestion
    ri_cur.truncate_to_canonical_form();

```

File: eyelid-match-ops/src/primitives/poly.rs [hunk 23]
```diff
@@ -29,3 +32,112 @@ pub mod test;
...
+    }
+
+    (x_prev, y_prev, ri_prev)
+}
+
+/// This sampling is similar to what will be necessary for YASHE KeyGen.
+/// The purpose is to obtain a polynomial with small random coefficients.
+pub fn sample<const MAX_POLY_DEGREE: usize>(mut rng: ThreadRng) -> Poly<MAX_POLY_DEGREE> {
+    let mut res = Poly::zero();
+    let max_coeff = 8;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly.rs:135]:
```suggestion
    // TODO: assert that this is less than the modulus of the coefficient
    let max_coeff = 6;
```
